lib/compress/zstd_double_fast.c

```diff
@@ -48,10 +48,216 @@ void ZSTD_fillDoubleHashTable(ZSTD_matchState_t* ms,
+    const BYTE* const ilimit = iend - HASH_READ_SIZE;
+    U32 offset_1=rep[0], offset_2=rep[1];
+    U32 offsetSaved = 0;
+
+    size_t mLength;
+    U32 offset;
+    U32 curr;
+
+    const size_t kStepIncr = 1 << kSearchStrength;
+    const BYTE* nextStep;
+    size_t step;
+
+    size_t hl0;
+    size_t hs0;
+    size_t hl1;
+
+    U32 idxl0;
+    U32 idxs0;
+    U32 idxl1;
+
+    const BYTE* matchl0;
+    const BYTE* matchs0;
+    const BYTE* matchl1;
+
+    const BYTE* ip = istart;
+    const BYTE* ip1;
+
+    DEBUGLOG(5, "ZSTD_compressBlock_doubleFast_noDict_generic");
+
+    /* init */
+    ip += ((ip - prefixLowest) == 0);
+    {
+        U32 const current = (U32)(ip - base);
+        U32 const windowLow = ZSTD_getLowestPrefixIndex(ms, current, cParams->windowLog);
+        U32 const maxRep = current - windowLow;
+        if (offset_2 > maxRep) offsetSaved = offset_2, offset_2 = 0;
+        if (offset_1 > maxRep) offsetSaved = offset_1, offset_1 = 0;
+    }
+
+_start:
+
+    step = 1;
+    nextStep = ip + kStepIncr;
+    ip1 = ip + step;
+
+    if (ip1 > ilimit) {
+        goto _cleanup;
+    }
+
+    hl0 = ZSTD_hashPtr(ip, hBitsL, 8);
+    idxl0 = hashLong[hl0];
+    matchl0 = base + idxl0;
+
+    /* Main Search Loop */
+    do {
+        curr = (U32)(ip-base);
+        hs0 = ZSTD_hashPtr(ip, hBitsS, mls);
+        idxs0 = hashSmall[hs0];
+        matchs0 = base + idxs0;
+
+        hashLong[hl0] = hashSmall[hs0] = curr;   /* update hash tables */
+
+        /* check noDict repcode */
+        if ((offset_1 > 0) & (MEM_read32(ip+1-offset_1) == MEM_read32(ip+1))) {
+            mLength = ZSTD_count(ip+1+4, ip+1+4-offset_1, iend) + 4;
+            ip++;
+            ZSTD_storeSeq(seqStore, (size_t)(ip-anchor), anchor, iend, 0, mLength-MINMATCH);
+            goto _match_stored;
+        }
+
+        hl1 = ZSTD_hashPtr(ip1, hBitsL, 8);
+
+        if (idxl0 > prefixLowestIndex) {
+            /* check prefix long match */
+            if (MEM_read64(matchl0) == MEM_read64(ip)) {
+                mLength = ZSTD_count(ip+8, matchl0+8, iend) + 8;
+                offset = (U32)(ip-matchl0);
+                while (((ip>anchor) & (matchl0>prefixLowest)) && (ip[-1] == matchl0[-1])) { ip--; matchl0--; mLength++; } /* catch up */
+                goto _match_found;
+            }
+        }
+
+        idxl1 = hashLong[hl1];
+        matchl1 = base + idxl1;
+
+        if (idxs0 > prefixLowestIndex) {
+            /* check prefix short match */
+            if (MEM_read32(matchs0) == MEM_read32(ip)) {
+                goto _search_next_long;
+            }
+        }
+
+        if (ip1 >= nextStep) {
+            PREFETCH_L1(ip1 + 64);
+            PREFETCH_L1(ip1 + 128);
+            step++;
+            nextStep += kStepIncr;
+        }
+        ip = ip1;
+        ip1 += step;
+
+        hl0 = hl1;
+        idxl0 = idxl1;
+        matchl0 = matchl1;
+#if defined(__aarch64__)
+        PREFETCH_L1(ip+256);
```

Comment:
Out of curiosity, why do we only prefetch here for ARM64?

Reply:
I dunno! I think @terrelln added that, this just preserves the existing behavior.